src/fields/mod.rs

```diff
@@ -0,0 +1,353 @@
+//! Various prime order fields. Each is defined in its own submodule because otherwise the names of
+//! items generated by the [`ff::PrimeField`] macro collide.
+use crate::Codec;
+use anyhow::{Context, anyhow};
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use std::{
+    fmt::Debug,
+    io::Cursor,
+    ops::{Add, AddAssign, Mul, Neg},
+};
+use subtle::{Choice, ConstantTimeEq};
+
+pub trait FieldElement:
+    Debug
+    + Clone
+    + Copy
+    + ConstantTimeEq
+    + From<u64>
+    + Add<Output = Self>
+    + for<'a> Add<&'a Self, Output = Self>
+    + AddAssign
+    + Mul<Output = Self>
+    + for<'a> Mul<&'a Self, Output = Self>
+    + Neg<Output = Self>
+    + for<'a> TryFrom<&'a [u8], Error = anyhow::Error>
+    + Codec
+{
+    const NUM_BITS: u32;
+    const ZERO: Self;
+    const ONE: Self;
+
+    /// Number of bytes needed to represent a field element.
+    fn num_bytes() -> usize {
+        (Self::NUM_BITS as usize).div_ceil(8)
+    }
+
+    /// Project an integer into the field.
+    fn from_u128(value: u128) -> Self;
+
+    /// Test whether this element is zero.
+    fn is_zero(&self) -> Choice {
```

Comment:
I used this method in a few places just because `ff` provided it, but I wonder if we should just use the more natural `v == FieldElement::ZERO`. I guess calling `is_zero` eventually lets us do the right thing w.r.t. constant time.

Reply:
I'm open to dropping this method, but I mainly wanted to make this PR a drop-in replacement.